azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java

```diff
@@ -0,0 +1,168 @@
+    } catch (final SchedulerException e) {
+      logger.error("Exception shutting down scheduler: ", e);
+    }
+  }
+
+  public void unregisterJob(final String groupName) throws SchedulerException {
+    if(!ifJobExist(groupName)) {
+      logger.warn("can not find job with " + groupName + " in quartz.");
+    } else {
+      this.scheduler.deleteJob(new JobKey("job1", groupName));
+    }
+  }
+
+  /**
+   * Only cron schedule register is supported.
+   *
+   * @param cronExpression the cron schedule for this job
+   * @param jobDescription Regarding QuartzJobDescription#groupName, in order to guarantee no
+   * duplicate quartz schedules, we design the naming convention depending on use cases:
+   * <ul>
+   *   <li>User flow schedule: we use {@link org.quartz.JobKey#JobKey} to represent the identity
+   *   of a flow's schedule. The format follows "$projectID_$flowName" to guarantee no duplicates.
+   *   </li>
+   *   <li>Quartz schedule for AZ internal use: the groupName should start with letters, rather
+   *   than number, which is the first case. </li>
+   * <ul>
+   */
+  public void registerJob(final String cronExpression, final QuartzJobDescription jobDescription)
+    throws SchedulerException {
+
+    requireNonNull(jobDescription, "jobDescription is null");
+
+    // Not allowed to register duplicate job name.
+    if(ifJobExist(jobDescription.getGroupName())) {
+      throw new SchedulerException("can not register existing job " + jobDescription.getGroupName());
+    }
+
+    if (!CronExpression.isValidExpression(cronExpression)) {
+      throw new SchedulerException("The cron expression string <" +  cronExpression + "> is not valid.");
+    }
+
+    // TODO kunkun-tang: Today all JobDetail should use "job1", and we will modify this when we
+    // start supporting multi schedules per flow.
+    final JobDetail job = JobBuilder.newJob(jobDescription.getJobClass())
+        .withIdentity("job1", jobDescription.getGroupName()).build();
+
+    // Add external dependencies to Job Data Map.
+    job.getJobDataMap().putAll(jobDescription.getContextMap());
+
+    // TODO kunkun-tang: Need management code to deal with different misfire policy
+    final Trigger trigger = TriggerBuilder
+        .newTrigger()
+        .withSchedule(
+            CronScheduleBuilder.cronSchedule(cronExpression)
+                .withMisfireHandlingInstructionFireAndProceed()
+//            .withMisfireHandlingInstructionDoNothing()
+//            .withMisfireHandlingInstructionIgnoreMisfires()
+        )
+        .build();
+
+    this.scheduler.scheduleJob(job, trigger);
+    logger.info("Quartz Schedule with jobDetail " + job.getDescription() + " is registered.");
+  }
+
+
+  public boolean ifJobExist(final String groupName) throws SchedulerException {
+    final Set<JobKey> jobKeySet = this.scheduler.getJobKeys(GroupMatcher.jobGroupEquals(groupName));
+    return jobKeySet !=null && jobKeySet.size() > 0;
+  }
+
+  Scheduler getScheduler() {
```

Comment:
access modifiers?

Reply:
Adding public modifier here. I will add a shut down hook with web server stop.

Reply:
Seems that shutdown doesn't need to `getScheduler`. For now, this method is never called by anyone.